Bearchem Subsystem Refactor

nullbear opened the pull request:
Refactors chemistry a bunch more.
- [x] Adds Chemistry Reaction Subsystem
- [x] Renames unclear variable. (mob_react to no_mob_react, as it should
  have been called.)
- [x] Prevents infinitely recursive recipes from freezing the game.
- [x] Logs potentially recursive recipes.
- [x] Prevents handle_reactions() from being called more than once per
  tick.
- [x] FUCKING DONE Get rid of snowflakey slime core checks.
- [x] Rework catalysts.
- [x] Add Ratio Recipe Logic.
- [ ] WIP: FUCK: Switch recipes over to new system.
- [ ] FUCK: switch reagents over to new system.
- [x] Split yucky atmos stuff into another branch.
- [x] Uses an associative list for holder.reagent_list, instead of iterating through each value and checking if the value.id == id that you want.

Essentially just maintains the chemistry module, and adds support for
new features to be added in the future. Probably raises the efficiency
of it a bunch too but idk how to profile that.

nullbear committed:
Bearchem Subsystem Refactor

Refactors chemistry a bunch more.

[x] Adds Chemistry Reaction Subsystem
[x] Renames unclear variable. (mob_react to no_mob_react, as it should
have been called.)
[x] Prevents infinitely recursive recipes from freezing the game.
[ ] Logs potentially recursive recipes.
[x] Prevents handle_reactions() from being called more than once per
tick.
[ ] WIP: Add support for reactions-over-time.
[ ] WIP: Get rid of snowflakey slime core checks.
[ ] WIP: Add support for advanced recipe requirements and reactions.
[ ] WIP: Rework catalysts.
[ ] WIP: Add Ratio Recipe Logic.
[ ] WIP: Switch recipes over to new system.

Essentially just maintains the chemistry module, and adds support for
new features to be added in the future. Probably raises the efficiency
of it a bunch too but idk how to profile that.

nullbear commented:
Slime core checks and container checks are being moved to the recipe, rather than being hard-coded into the holder. Reaction code is also being moved to recipes for the same reason.

F-OS commented:
i am not a coder, just a regular player, how will this affect me?

Supermichael777 commented:
you should notice almost no real changes, other then that you can mix smaller units of chems

nullbear commented:
Yeah that^ this pr should have virtually no effect on game play. And just serves as a framework for new features and better code efficiency/reuse ability.

nullbear commented:
I do however have 3 related prs that I intend to tie into this, which will have game play effects.

nullbear commented:
So ideally as a player, you might notice less lag?

Razharas commented:
Would be cool if we'd get rid of ids and switch to proper types which are simply better
Would be cool if we'd replaced the reagent list being a list of datums with list of lists with organised needed parametres like in listmos

nullbear commented:
types are awful for readability compared to the id's. But recipes have no real usage for ids and im getting rid of those, but keeping reagent id's.

I've tried a number of different methods for handling chemistry with/without id's, and honestly there is no substantial benefit to using types over id's. More notably, it reduces readability // thus reducing coding speed // and changes a bunch of stuff for no real reason except to change it. 

list of lists could be quite efficient, but im not putting in the hours of work needed to transfer everything over to use that mechanic, as most of chemistry would need to be rewritten to support it. Besides, it essentially IS a list of lists right now anyways, seeing as var/vars is what you're accessing to get at the variables.

Where listmos is gas_mixture[oxygen][moles] this is reagent_list[oxygen][volume], the syntax is just different, but mechanically it's the same.

I repeat, chemistry is set up USING TYPES. It just has a fuckton of helper procs to aid with code readability and qol, so youre used to using id's because its faster and easier, but you're welcome to do it the hard way via types, it's just kind of a waste of time, and the helper procs are made to streamline the process anyways.

nullbear commented:
Literally looking at the vars that reagents use, and the vars that are accessed by other files.

What the heck do you want to 'listmos'-ize? There are like 4 values that could potentially be useful like that, and there's really no reason to want to do that except to make it consisten with atmos. 

id // which is being gotten rid of in your ideal vision for reagents anyways.
volume //
data //
color // 

I see no real benefit of `reagent_list["oxygen"]["volume"]` over `reagent_list["oxygen"].volume` and if we were getting rid of id's, as so many people want to do. The main reason a list of lists is desired is lost anyways? Because it becomes `reagent_list[/datum/reagent/oxygen]["volume"]`

nullbear commented:
The only other useful reason to use types over id's is so that in recipes, you can accept subtypes as acceptable substitutes. Such as a recipe calling for ethanol, you can use any kind of prepared ethanol drink as a substitute as well. 

EXCEPT THIS LOGIC IS BEING IMPLEMENTED IN THIS REFACTOR ANYWAYS, AND DOESNT REQUIRE ID'S TO BE REMOVED.

There is a 'special_reqs' portion of recipes in this PR which allow you to use non-standard recipe requirements. From requiring a specific container like slime cores, to requiring that the reaction takes place on a certain z-level or with freon in the environment etc. etc. You can use the special reqs to do the same subtyping thing that you would potentially want to use types for over id's. Except the bonus is that it doesnt mean you NEED to use the subtyping for all recipes, only the recipes that you actually want to use it for.

And im sorry for the long ass rant, and i dont mean to target you in particular because virtually everyone who has glanced at chemistry code (including me) has thought and said at some point 'why do we even use id's? We should use types.'

Which is why im making all of these clarifications. I've worked with chemcode for quite a bit, and there is no reason to get rid of id's, as it would just reduce code readability, and slow down development. This code refactor allows using all the benefits of type/datum/list reagents, and more, without reducing code readability in the process.

Razharas commented:
Yes having list of lists is simply for potential transformability to atmos
Chemmos is inevitable although perhaps far future

nullbear commented:
It would make more sense to merge gas_datum and reagents together, and only use the variables you need for each system. IE 

datum/material/oxygen
  specific heat
  id
  volume
  moles
 etc

and when it's used by atmos, it uses moles + specific heat + id. When its used by chemistry it uses id + volume. You don't need to have chemistry using all the same mechanics as atmos does, especially not right away, as that's just unnecessary complexity.

nullbear committed:
Starts porting slime extracts to new reaction system.

Feedback on this one maybe? It's about the same as current reactions in
terms of implementation. But much more dynamic since the recipe actually
handles the recipe, rather than recipe handling being hardcoded into
holder. Allowing for recipes to be more dynamic than just 'mix X Y for
Z'

nullbear commented:
I've got a funeral to go to and wont be home for a couple days.

nullbear committed:
WIP commit 2

Adds simpler reagent processing proc. "adjust_volume" which quickly adds
or removes a reagent to the holder. It is used only when creating or
deleting reagents, instead of transferring, and currently is incomplete.

Adds a bunch of modular reaction procs that will be used frequently. It
is essentially just normal reaction code broken down to be more
reuseable and configureable.

Ports all of the slime code to use this system. Properly removing the
slimecode handling from holder.dm. I fucked with the naming scheme of
things a bit and it's going to gib stats gathering. @korphaeron i figure
you know better who to talk to about this.

nullbear commented:
@korphaeron this breaks stats gathering. Who do i collab w/ to make this okay again?

nullbear committed:
WIP Commit 3

starts working on at48 defines

gets rid of a few niche holder procs, replacing them with faster and
easier to use ones which are much more reuseable.

Makes reagents_list use an associated list by reagent id instead.

Adds support for setting starting reagents on new.

Adds Adjust_Volume proc. The coverall add/remove reagent proc for
creating and deleting reagents. Simply set the ID and Volume Delta that
you want to change. Negative Values delete, Positive values create.

Maintains remove_amount. It removes an amount of units from the
containers, keeping the average ratios of each chemical the same.

Get_Master_Reagent is added to reduce copypasta.

trans_to is replaced with transfer proc. It transfers an amount from
this holder to target holder using much faster procs.

Copy_to proc is maintained.

Transfer_id replaces trans_id_to. As does copy_id_to.

Removes a couple of redundant safety checks from metabolize, and marks
it for removal at a future date.

Finallizes work to prevent handle_reactions() from being called more
than once per tick. Very happy with this.

Set_reacting is used to mark the holder for reacting. It reacts once
this tick, and then again every two ticks for as long as its reacting.

Adds support for new reaction feature. Currently unused.

Adds logging of potentially recursive recipes.

Removes snowflake checks. Eventually will move them to their respective
reagents.

Maintains update_total(). It's faster, and rounds the total to the
nearest MIN_REAGENT_VOLUME, which is currently set at 5 milliunits. This
prevents REALLY long unit decimal placements.

Removes nearly identical niche procs and replaces them with one
get_volume proc. This can be used either to check if a reagent exists,
and returns the volume of that reagent. Changed the return to be -1
instead of 0, so that zero can be used as an acceptible recipe
requirement.

nullbear committed:
Starts transfering recipes and adding RID_defines

KorPhaeron commented:
pm liamlime on tg forums for stats

nullbear committed:
shifts some reagents

nullbear committed:
commit 6

Razharas commented:
Yeye at least making it an associated list is better than nothing

nullbear commented:
I'm gunna split this into two PR's._

phil235 commented on code/modules/reagents/chemistry/defines.dm:
```diff
@@ -0,0 +1,90 @@
+#define RID_amasec		"amasec"
+#define RID_andalusia	"andalusia"
+#define RID_antifreeze	"antifreeze"
+#define RID_b52			"bf2"
```
I think you mean "b52"

nullbear replied:
ty

nullbear committed:
Unrelated WIP: Do not look at what this does.

nullbear committed:
Adds a temporary readme.

optimumtact commented:
I looked at your unrelated WIP, I saw the truth

phil235 commented on code/__DEFINES/materials/basic.dm:
```diff
@@ -0,0 +1,340 @@
+// Basic and tier-0 materials. These are typically of the lowest value, and the simplest/purest elements.
+
+#define MAT_hydro "h2"
+/datum/mat/hydrogen
```
The DEFINES folder is used solely to store defines, not to store new datum or objects.

nullbear replied:
These datums are never instantiated.

I have two options, 1: Make a separate file, one with a bunch of defines, and one with a bunch of uninstantiated datum defines. This makes it so that if you want to find the ID of a datum, you have to go BACK to the defines folder and skim through it until you find the proper define. Or alternatively, skim through the defines and then have to use the search function to find out where its actually used, so that you know what it is.

The other option is this: to define them WITH the datums. But the issue there is that it needs to be defined before it will be used in the code. IE. before the game or module folders.

nullbear commented:
oranges, i intend to split it into two PR's.

nullbear committed:
Last WIP commit before split.

nullbear committed:
Splits into two PR's

nullbear committed:
oops forgot to remove these

nullbear committed:
r

MrStonedOne commented:
>I see no real benefit of `reagent_list["oxygen"]["volume"]` over `reagent_list["oxygen"].volume`

The main one is speed. note, atmos isn't "volume", its VOLUME, and that resolves to an index. accessing lists by index are O(1), accessing datum vars is O(log(n)) since datum vars are always accessed by name, and thats 511 only. 510 its O(n), meaning even associated var access (O(log(n))) is faster.

If you ever look at how atmos does things, and think that it's doing it an odd way, speed is almost always why.

nullbear commented:
Yea, MrStonedOne, I've started on a separate working branch where i'm experimenting with reagents and chemistry using the same system/design as listmos. 

The 'no real benefit' thing was because reagents are instantiated, and without making a completely different system for them (the experimental branch) there would be next to no benefit to reagent_list[][] over reagent_list[].vol because it's still just being instantiated.

nullbear committed:
so much shit ugh

phil235 commented on code/modules/reagents/chemistry/holder.dm:
```diff
@@ -1,665 +0,0 @@
-//This file was auto-corrected by findeclaration.exe on 25.5.2012 20:42:32
```
Why didn't you move all these lines to a new file just to change the file's name? It makes it very hard to spot the changes you've made in these lines.

phil235 commented on code/modules/reagents/chemistry/readme.dm:
```diff
@@ -6,7 +6,7 @@ Structure: ///////////////////          //////////////////////////
 		   			|				    			 |
     The object that holds everything.   			 V
-		   							      reagent_list var (list)   	A List of datums, each datum is a reagent.
+		   							      reagents var (list)   	A List of datums, each datum is a reagent.
```
Wait, so now /datum/reagents has a variable that is named "reagents" ?????

phil235 commented on code/modules/reagents/chemistry/chem_holder.dm:
```diff
@@ -0,0 +1,524 @@
+///////////////////////////////////////////////////////////////////////////////////
+
+/datum/reagents
+	var/list/reagents = new/list()
```
/datum/reagents has a var named reagents?

nullbear replied:
datum/reagents should really be datum/reagent_holder or something, but it'd be hell to change.

phil235 replied:
agreed.

nullbear commented:
OH THATS NEAT. The github website now shows which files have merge conflicts!!

Anyways im closing this, as most of my work has been redone better in another branch.

Razharas commented:
Ok

nullbear commented:
https://github.com/nullbear/-tg-station/tree/fuck-why-am-i-starting-so-many-huge-projects-and-never-finishing-them-for-fucks-sakes-help-me- 

Is my current working branch for this if anyone wants to help out/check it out/make suggestions while im working on it. I'm not going to make a PR for it until its actually nearing completion though.
